File: app/src/main/java/com/belous/v/clrc/other/di/AppModule.java
package com.belous.v.clrc.other.di;

import android.content.Context;

import androidx.room.Room;

import com.belous.v.clrc.model.db.AppDatabase;
import com.belous.v.clrc.model.db.ItemDao;
import com.belous.v.clrc.presenter.InfoPresenter;
import com.belous.v.clrc.presenter.InfoPresenterImpl;
import com.belous.v.clrc.presenter.ListPresenter;
import com.belous.v.clrc.presenter.ListPresenterImpl;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
class AppModule {

    private Context context;

    AppModule(Context context) {
        this.context = context;
    }

    @Singleton
    @Provides
    Context getContext() {
        return context;
    }

    @Singleton
    @Provides
    AppDatabase provideAppDatabase(Context context) {
        return Room.databaseBuilder(context, AppDatabase.class, AppDatabase.DB_NAME).build();
    }

    @Singleton
    @Provides
    ItemDao provideItemDao(AppDatabase appDatabase) {
        return appDatabase.itemDao();
    }

    @Singleton
    @Provides
    ListPresenter provideListPresenter(ItemDao itemDao) {
        return new ListPresenterImpl(itemDao);
    }

    @Singleton
    @Provides
    InfoPresenter provideInfoPresenter(ItemDao itemDao) {
        return new InfoPresenterImpl(itemDao);
    }
}

File: app/src/main/java/com/belous/v/clrc/di/AppModule.kt
package com.belous.v.clrc.di

import android.app.Application
import androidx.room.Room
import com.belous.v.clrc.MainStates
import com.belous.v.clrc.data.db.AppDatabase
import com.belous.v.clrc.data.db.YeelightDao
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
object AppModule {

    @Singleton
    @Provides
    fun provideMainStates() = MainStates()

    @Singleton
    @Provides
    fun provideAppDatabase(application: Application): AppDatabase {
        return Room.databaseBuilder(
            application,
            AppDatabase::class.java,
            AppDatabase.DB_NAME
        ).build()
    }

    @Singleton
    @Provides
    fun provideItemDao(appDatabase: AppDatabase): YeelightDao {
        return appDatabase.itemDao()
    }
}